File: typescript/tests/consent.test.ts
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

import { beforeEach, describe, expect, test, vi } from 'vitest';

import {
  type ConsentLevel,
  getConsentLevel,
  loadConsentFromEnv,
  resetConsentForTests,
  setConsentLevel,
  shouldAllow,
} from '../src/consent.js';

beforeEach(() => {
  resetConsentForTests();
  delete process.env['PROVIDE_CONSENT_LEVEL'];
});

describe('ConsentLevel defaults', () => {
  test('default consent is FULL', () => {
    expect(getConsentLevel()).toBe('FULL');
  });
});

describe('ConsentLevel.FULL', () => {
  beforeEach(() => setConsentLevel('FULL'));

  test('allows logs at any level', () => {
    expect(shouldAllow('logs', 'DEBUG')).toBe(true);
    expect(shouldAllow('logs', 'INFO')).toBe(true);
    expect(shouldAllow('logs', 'WARNING')).toBe(true);
    expect(shouldAllow('logs', 'ERROR')).toBe(true);
  });

  test('allows traces', () => {
    expect(shouldAllow('traces')).toBe(true);
  });

  test('allows metrics', () => {
    expect(shouldAllow('metrics')).toBe(true);
  });

  test('allows context', () => {
    expect(shouldAllow('context')).toBe(true);
  });
});

describe('ConsentLevel.NONE', () => {
  beforeEach(() => setConsentLevel('NONE'));

  test('blocks all signals', () => {
    expect(shouldAllow('logs', 'ERROR')).toBe(false);
    expect(shouldAllow('traces')).toBe(false);
    expect(shouldAllow('metrics')).toBe(false);
    expect(shouldAllow('context')).toBe(false);
  });
});

describe('ConsentLevel.FUNCTIONAL', () => {
  beforeEach(() => setConsentLevel('FUNCTIONAL'));

  test('blocks logs below WARNING', () => {
    expect(shouldAllow('logs', 'DEBUG')).toBe(false);
    expect(shouldAllow('logs', 'INFO')).toBe(false);
  });

  test('allows logs at WARNING and above', () => {
    expect(shouldAllow('logs', 'WARNING')).toBe(true);
    expect(shouldAllow('logs', 'ERROR')).toBe(true);
    expect(shouldAllow('logs', 'CRITICAL')).toBe(true);
  });

  test('allows logs at WARN (alias)', () => {
    expect(shouldAllow('logs', 'WARN')).toBe(true);
  });

  test('allows traces', () => {
    expect(shouldAllow('traces')).toBe(true);
  });

  test('allows metrics', () => {
    expect(shouldAllow('metrics')).toBe(true);
  });

  test('blocks context', () => {
    expect(shouldAllow('context')).toBe(false);
  });

  test('allows unknown signals', () => {
    expect(shouldAllow('custom_signal')).toBe(true);
  });

  test('blocks logs with no log level', () => {
    expect(shouldAllow('logs')).toBe(false);
    expect(shouldAllow('logs', undefined)).toBe(false);
  });
});

describe('ConsentLevel.MINIMAL', () => {
  beforeEach(() => setConsentLevel('MINIMAL'));

  test('blocks logs below ERROR', () => {
    expect(shouldAllow('logs', 'DEBUG')).toBe(false);
    expect(shouldAllow('logs', 'INFO')).toBe(false);
    expect(shouldAllow('logs', 'WARNING')).toBe(false);
  });

  test('allows logs at ERROR and above', () => {
    expect(shouldAllow('logs', 'ERROR')).toBe(true);
    expect(shouldAllow('logs', 'CRITICAL')).toBe(true);
  });

  test('blocks traces even with ERROR-level logLevel argument', () => {
    expect(shouldAllow('traces', 'ERROR')).toBe(false);
  });

  test('blocks traces', () => {
    expect(shouldAllow('traces')).toBe(false);
  });

  test('blocks metrics even with ERROR-level logLevel argument', () => {
    expect(shouldAllow('metrics', 'ERROR')).toBe(false);
  });

  test('blocks metrics', () => {
    expect(shouldAllow('metrics')).toBe(false);
  });

  test('blocks context', () => {
    expect(shouldAllow('context')).toBe(false);
  });

  test('blocks unknown signals', () => {
    expect(shouldAllow('custom_signal')).toBe(false);
  });

  test('blocks logs with no log level', () => {
    expect(shouldAllow('logs')).toBe(false);
    expect(shouldAllow('logs', undefined)).toBe(false);
  });
});

describe('setConsentLevel / getConsentLevel', () => {
  test('set and get round-trip', () => {
    const levels: ConsentLevel[] = ['FULL', 'FUNCTIONAL', 'MINIMAL', 'NONE'];
    for (const level of levels) {
      setConsentLevel(level);
      expect(getConsentLevel()).toBe(level);
    }
  });
});

describe('loadConsentFromEnv', () => {
  test('loads FULL from env (overriding non-FULL state)', () => {
    setConsentLevel('NONE');
    process.env['PROVIDE_CONSENT_LEVEL'] = 'FULL';
    loadConsentFromEnv();
    expect(getConsentLevel()).toBe('FULL');
  });

  test('loads FUNCTIONAL from env', () => {
    process.env['PROVIDE_CONSENT_LEVEL'] = 'FUNCTIONAL';
    loadConsentFromEnv();
    expect(getConsentLevel()).toBe('FUNCTIONAL');
  });

  test('loads MINIMAL from env', () => {
    process.env['PROVIDE_CONSENT_LEVEL'] = 'MINIMAL';
    loadConsentFromEnv();
    expect(getConsentLevel()).toBe('MINIMAL');
  });

  test('loads NONE from env', () => {
    process.env['PROVIDE_CONSENT_LEVEL'] = 'NONE';
    loadConsentFromEnv();
    expect(getConsentLevel()).toBe('NONE');
  });

  test('every valid value applies over any prior level', () => {
    const levels: ConsentLevel[] = ['FULL', 'FUNCTIONAL', 'MINIMAL', 'NONE'];
    for (const level of levels) {
      setConsentLevel(level === 'FULL' ? 'NONE' : 'FULL');
      process.env['PROVIDE_CONSENT_LEVEL'] = level;
      loadConsentFromEnv();
      expect(getConsentLevel()).toBe(level);
    }
  });

  test('an invalid env value fails closed to NONE, overriding the current level', () => {
    const warnSpy = vi.spyOn(console, 'warn').mockImplementation(() => undefined);
    try {
      setConsentLevel('MINIMAL');
      process.env['PROVIDE_CONSENT_LEVEL'] = 'BOGUS';
      loadConsentFromEnv();
      expect(getConsentLevel()).toBe('NONE');
      expect(warnSpy).toHaveBeenCalledTimes(1);
    } finally {
      warnSpy.mockRestore();
    }
  });

  test('leaves the default FULL in place when env is unset', () => {
    delete process.env['PROVIDE_CONSENT_LEVEL'];
    loadConsentFromEnv();
    expect(getConsentLevel()).toBe('FULL');
  });

  test('unset env leaves a programmatic level untouched (no reset to FULL)', () => {
    setConsentLevel('MINIMAL');
    delete process.env['PROVIDE_CONSENT_LEVEL'];
    loadConsentFromEnv();
    expect(getConsentLevel()).toBe('MINIMAL');
  });

  test('empty env value is blank, not invalid, and leaves the level untouched', () => {
    setConsentLevel('FUNCTIONAL');
    process.env['PROVIDE_CONSENT_LEVEL'] = '';
    loadConsentFromEnv();
    expect(getConsentLevel()).toBe('FUNCTIONAL');
  });

  test('handles lowercase env value', () => {
    process.env['PROVIDE_CONSENT_LEVEL'] = 'minimal';
    loadConsentFromEnv();
    expect(getConsentLevel()).toBe('MINIMAL');
  });

  test('handles env value with whitespace', () => {
    process.env['PROVIDE_CONSENT_LEVEL'] = '  NONE  ';
    loadConsentFromEnv();
    expect(getConsentLevel()).toBe('NONE');
  });

  test('reads nothing and leaves the level untouched when process is unavailable', () => {
    setConsentLevel('MINIMAL');
    process.env['PROVIDE_CONSENT_LEVEL'] = 'NONE';
    const realProcess = globalThis.process;
    // Simulate a browser bundle: `process` is not defined at all.
    vi.stubGlobal('process', undefined);
    try {
      loadConsentFromEnv();
      expect(getConsentLevel()).toBe('MINIMAL');
    } finally {
      vi.stubGlobal('process', realProcess);
      vi.unstubAllGlobals();
    }
  });

  test('swallows an env accessor that throws and leaves the level untouched', () => {
    setConsentLevel('MINIMAL');
    const realProcess = globalThis.process;
    const throwing = new Proxy(
      {},
      {
        get: () => {
          throw new Error('env unavailable');
        },
      },
    );
    vi.stubGlobal('process', { env: throwing });
    try {
      loadConsentFromEnv();
      expect(getConsentLevel()).toBe('MINIMAL');
    } finally {
      vi.stubGlobal('process', realProcess);
      vi.unstubAllGlobals();
    }
  });
});

File: go/consent_test.go
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0
// SPDX-Comment: Part of provide-telemetry.

package telemetry

import (
	"context"
	"os"
	"testing"
)

func TestConsentDefaultIsFull(t *testing.T) {
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)
	if got := GetConsentLevel(); got != ConsentFull {
		t.Errorf("expected ConsentFull, got %v", got)
	}
}

func TestConsentFullAllowsAllSignals(t *testing.T) {
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)
	SetConsentLevel(ConsentFull)
	signals := []struct {
		signal   string
		logLevel string
	}{
		{"logs", "DEBUG"},
		{"traces", ""},
		{"metrics", ""},
		{"context", ""},
	}
	for _, s := range signals {
		if !ShouldAllow(s.signal, s.logLevel) {
			t.Errorf("expected FULL to allow %s", s.signal)
		}
	}
}

func TestConsentNoneBlocksAllSignals(t *testing.T) {
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)
	SetConsentLevel(ConsentNone)
	signals := []struct {
		signal   string
		logLevel string
	}{
		{"logs", "ERROR"},
		{"traces", ""},
		{"metrics", ""},
		{"context", ""},
	}
	for _, s := range signals {
		if ShouldAllow(s.signal, s.logLevel) {
			t.Errorf("expected NONE to block %s", s.signal)
		}
	}
}

func TestConsentFunctionalLogThresholds(t *testing.T) {
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)
	SetConsentLevel(ConsentFunctional)

	tests := []struct {
		logLevel string
		want     bool
	}{
		{"DEBUG", false},
		{"INFO", false},
		{"WARNING", true},
		{"WARN", true},
		{"ERROR", true},
		{"CRITICAL", true},
		{"", false},
	}
	for _, tt := range tests {
		got := ShouldAllow("logs", tt.logLevel)
		if got != tt.want {
			t.Errorf("FUNCTIONAL logs %q: got %v, want %v", tt.logLevel, got, tt.want)
		}
	}
}

func TestConsentFunctionalTracesAndMetricsAllowed(t *testing.T) {
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)
	SetConsentLevel(ConsentFunctional)
	if !ShouldAllow("traces", "") {
		t.Error("expected FUNCTIONAL to allow traces")
	}
	if !ShouldAllow("metrics", "") {
		t.Error("expected FUNCTIONAL to allow metrics")
	}
}

func TestConsentFunctionalContextBlocked(t *testing.T) {
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)
	SetConsentLevel(ConsentFunctional)
	if ShouldAllow("context", "") {
		t.Error("expected FUNCTIONAL to block context")
	}
}

func TestConsentMinimalLogThresholds(t *testing.T) {
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)
	SetConsentLevel(ConsentMinimal)

	tests := []struct {
		logLevel string
		want     bool
	}{
		{"DEBUG", false},
		{"INFO", false},
		{"WARNING", false},
		{"ERROR", true},
		{"CRITICAL", true},
		{"", false},
	}
	for _, tt := range tests {
		got := ShouldAllow("logs", tt.logLevel)
		if got != tt.want {
			t.Errorf("MINIMAL logs %q: got %v, want %v", tt.logLevel, got, tt.want)
		}
	}
}

func TestConsentMinimalBlocksTracesMetricsContext(t *testing.T) {
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)
	SetConsentLevel(ConsentMinimal)
	for _, sig := range []string{"traces", "metrics", "context"} {
		if ShouldAllow(sig, "") {
			t.Errorf("expected MINIMAL to block %s", sig)
		}
	}
}

func TestLoadConsentFromEnvFull(t *testing.T) {
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)
	t.Setenv("PROVIDE_CONSENT_LEVEL", "FULL")
	LoadConsentFromEnv()
	if got := GetConsentLevel(); got != ConsentFull {
		t.Errorf("expected ConsentFull, got %v", got)
	}
}

func TestLoadConsentFromEnvFunctional(t *testing.T) {
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)
	t.Setenv("PROVIDE_CONSENT_LEVEL", "FUNCTIONAL")
	LoadConsentFromEnv()
	if got := GetConsentLevel(); got != ConsentFunctional {
		t.Errorf("expected ConsentFunctional, got %v", got)
	}
}

func TestLoadConsentFromEnvMinimal(t *testing.T) {
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)
	t.Setenv("PROVIDE_CONSENT_LEVEL", "MINIMAL")
	LoadConsentFromEnv()
	if got := GetConsentLevel(); got != ConsentMinimal {
		t.Errorf("expected ConsentMinimal, got %v", got)
	}
}

func TestLoadConsentFromEnvNone(t *testing.T) {
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)
	t.Setenv("PROVIDE_CONSENT_LEVEL", "NONE")
	LoadConsentFromEnv()
	if got := GetConsentLevel(); got != ConsentNone {
		t.Errorf("expected ConsentNone, got %v", got)
	}
}

func TestLoadConsentFromEnvInvalidIgnored(t *testing.T) {
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)
	t.Setenv("PROVIDE_CONSENT_LEVEL", "BOGUS")
	LoadConsentFromEnv()
	// invalid value leaves level unchanged (FULL)
	if got := GetConsentLevel(); got != ConsentFull {
		t.Errorf("expected ConsentFull after bogus value, got %v", got)
	}
}

func TestLoadConsentFromEnvEmpty(t *testing.T) {
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)
	t.Setenv("PROVIDE_CONSENT_LEVEL", "")
	LoadConsentFromEnv()
	// empty env var leaves level unchanged
	if got := GetConsentLevel(); got != ConsentFull {
		t.Errorf("expected ConsentFull after empty env, got %v", got)
	}
}

// ── Environment wiring ────────────────────────────────────────────────────────
//
// LoadConsentFromEnv existed from the start; nothing called it, so
// PROVIDE_CONSENT_LEVEL=NONE was read by every other SDK and ignored by Go.
// These pin the two entry points that now load it — setup and the lazy
// pre-setup logger — and the two things they must not do: overwrite a level
// when the variable is unset, and overwrite a programmatic level after setup.

func TestSetupTelemetryLoadsConsentFromEnv(t *testing.T) {
	resetSetupState(t)
	t.Cleanup(func() { resetSetupState(t) })
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)

	t.Setenv("PROVIDE_CONSENT_LEVEL", "NONE")
	if _, err := SetupTelemetry(); err != nil {
		t.Fatalf("setup failed: %v", err)
	}
	if got := GetConsentLevel(); got != ConsentNone {
		t.Fatalf("expected setup to load ConsentNone from env, got %v", got)
	}
	if ShouldAllow(signalLogs, "ERROR") {
		t.Fatal("expected NONE loaded at setup to block logs")
	}
}

func TestSetupTelemetryLeavesConsentWhenEnvUnset(t *testing.T) {
	resetSetupState(t)
	t.Cleanup(func() { resetSetupState(t) })
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)

	t.Setenv("PROVIDE_CONSENT_LEVEL", "")
	_ = os.Unsetenv("PROVIDE_CONSENT_LEVEL")
	SetConsentLevel(ConsentMinimal)
	if _, err := SetupTelemetry(); err != nil {
		t.Fatalf("setup failed: %v", err)
	}
	if got := GetConsentLevel(); got != ConsentMinimal {
		t.Fatalf("expected unset env to leave ConsentMinimal in place, got %v", got)
	}
}

func TestGetLoggerBeforeSetupLoadsConsentFromEnv(t *testing.T) {
	resetSetupState(t)
	t.Cleanup(func() { resetSetupState(t) })
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)

	t.Setenv("PROVIDE_CONSENT_LEVEL", "NONE")
	before := GetHealthSnapshot()
	GetLogger(context.Background(), "x").Info("lazy.logger.consent.none")
	if got := GetConsentLevel(); got != ConsentNone {
		t.Fatalf("expected lazy GetLogger to load ConsentNone from env, got %v", got)
	}
	if after := GetHealthSnapshot(); after.LogsEmitted != before.LogsEmitted {
		t.Fatalf("expected NONE to suppress the lazy record: before=%d after=%d", before.LogsEmitted, after.LogsEmitted)
	}
}

func TestGetLoggerAfterSetupDoesNotReloadConsentFromEnv(t *testing.T) {
	resetSetupState(t)
	t.Cleanup(func() { resetSetupState(t) })
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)

	t.Setenv("PROVIDE_CONSENT_LEVEL", "NONE")
	if _, err := SetupTelemetry(); err != nil {
		t.Fatalf("setup failed: %v", err)
	}
	// A programmatic decision made after setup is the authority; a later
	// GetLogger must not clobber it with the environment again.
	SetConsentLevel(ConsentFull)
	GetLogger(context.Background(), "x").Info("post.setup.consent.full")
	if got := GetConsentLevel(); got != ConsentFull {
		t.Fatalf("expected GetLogger after setup to leave ConsentFull, got %v", got)
	}
}

func TestConsentFunctionalUnknownSignalAllowed(t *testing.T) {
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)
	SetConsentLevel(ConsentFunctional)
	if !ShouldAllow("custom_signal", "") {
		t.Error("expected FUNCTIONAL to allow unknown signals")
	}
}

func TestConsentMinimalUnknownSignalBlocked(t *testing.T) {
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)
	SetConsentLevel(ConsentMinimal)
	if ShouldAllow("custom_signal", "") {
		t.Error("expected MINIMAL to block unknown signals")
	}
}

func TestSetGetConsentLevel(t *testing.T) {
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)
	SetConsentLevel(ConsentFunctional)
	if got := GetConsentLevel(); got != ConsentFunctional {
		t.Errorf("expected ConsentFunctional, got %v", got)
	}
}

func TestShouldAllowUnknownConsentLevelReturnsFalse(t *testing.T) {
	ResetConsentForTests()
	t.Cleanup(ResetConsentForTests)
	SetConsentLevel(ConsentLevel(99))
	if ShouldAllow("logs", "INFO") {
		t.Error("expected unknown ConsentLevel to deny all signals")
	}
}
